src/NLog/IMessageTemplate.cs

```diff
@@ -0,0 +1,144 @@
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+namespace NLog
+{
+    using System.Collections;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// 
+    /// </summary>
+    public struct MessageTemplateParameter
+    {
+        /// <summary>
+        /// 
+        /// </summary>
+        public readonly string Name;
+        /// <summary>
+        /// 
+        /// </summary>
+        public readonly object Value;
+        /// <summary>
+        /// 
+        /// </summary>
+        public readonly string Format;
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="name"></param>
+        /// <param name="value"></param>
+        /// <param name="format"></param>
+        public MessageTemplateParameter(string name, object value, string format)
+        {
+            Name = name;
+            Value = value;
+            Format = format;
+        }
+    }
+
+    /// <summary>
+    /// 
+    /// </summary>
+    public interface IMessageTemplateParameterList : IEnumerable<MessageTemplateParameter>
+    {
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="index"></param>
+        /// <returns></returns>
+        MessageTemplateParameter this[int index] { get; }
+        /// <summary>
+        /// 
+        /// </summary>
+        int Count { get; }
+    }
+
+    /// <summary>
+    /// 
+    /// </summary>
+    public interface IMessageTemplate
+    {
+        /// <summary>
+        /// 
+        /// </summary>
+        string Message { get; }
+        /// <summary>
+        /// 
+        /// </summary>
+        bool IsPositional { get; }
+        /// <summary>
+        /// 
+        /// </summary>
+        IMessageTemplateParameterList Parameters { get; }
+    }
+
+    internal class MessageTemplate : IMessageTemplate, IMessageTemplateParameterList
+    {
+        /// <summary>
+        /// 
+        /// </summary>
+        public string Message { get; }
+        /// <summary>
+        /// 
+        /// </summary>
+        public bool IsPositional { get; }
+        /// <summary>
+        /// 
+        /// </summary>
+        public IMessageTemplateParameterList Parameters { get { return this; } }
```

Comment:
Is this for performance reasons? It's a bit strange design to have a property that is the same as `this` IMO

Reply:
It is to minimize allocations and because we are not using IReadOnlyList